README.md
# realesrgan-scheduler

[![realesrgan-scheduler](https://github.com/kmulvey/realesrgan-scheduler/actions/workflows/release_build.yml/badge.svg)](https://github.com/kmulvey/realesrgan-scheduler/actions/workflows/release_build.yml) [![codecov](https://codecov.io/gh/kmulvey/realesrgan-scheduler/branch/main/graph/badge.svg?token=XpJ5kCJzsn)](https://codecov.io/gh/kmulvey/realesrgan-scheduler) [![Go Report Card](https://goreportcard.com/badge/github.com/kmulvey/realesrgan-scheduler)](https://goreportcard.com/report/github.com/kmulvey/realesrgan-scheduler) [![Go Reference](https://pkg.go.dev/badge/github.com/kmulvey/realesrgan-scheduler.svg)](https://pkg.go.dev/github.com/kmulvey/realesrgan-scheduler)


```
./realesrgan-scheduler-auto \
-realesrgan-path ~/bin/realesrgan-ncnn-vulkan-20220424-ubuntu/realesrgan-ncnn-vulkan \
-original-images-dir ~/Photos/ \
-upscaled-images-dir ~/upscayl/ \
-cache-dir ./skipcache/ \
-skip-file ../../largeskiplist.txt
```


CHANGELOG.md

0.1.0
=============
2023-04-05

* badges (0da728d5)
* bump deps (3babbcff)
* bump deps (466d7ba0)
* fix windows ci cd (f3b5e772)
* deps (ab4aa2eb)
* normalize ci cd configs (c83b40c1)
* bail on error (4a6d39c5)
* remove http (5330bfa7)
* remove old relase target (1f5e58d1)
* stop sending me emails (86fd4ffc)
* deps (42ca2e4f)
* deps (cf8a3be0)
* remove old build dir (09007336)
* new path lib (c25b032f)
* deps (48174247)
* imports (c1112ce8)
* upgrade path lib (81aba1f3)
* merge cleanup and compare (e5179783)
* ignore dist (09bc4ec1)
* fix releaser config (ee6724e5)
* linting (e956ba49)
* parallel (b5e5bf72)
* consolidate pretty printing (e55870cc)
* binary paths (ff679ec1)
* lint and release configs (2f757881)
* normalize action (14fdee3c)
* rename http pkg (8c31d5e1)
* linting and github action (6913a799)
* remove client (65c63713)
* show correct image sizes (24ed1d41)
* cleanup and compare cmds (44010fb5)
* build dirs (c6d6e584)
* more ignore (020203f5)
* ignore binaries (546011d3)
* support ignore lists (68caf476)
* remove extra verbose build args (f34e033d)
* rm (ea3069d0)
* dont bother if the len is 0 (718a6541)
* pre loop logging (412b27a8)
* add to cache (37c32538)
* add logging for non-watch (c5d30d7f)
* comments (c76dc63a)
* remaining queue length (8f7faf07)
* move logging to app from worker to get more accurate queue lenghts (90ef1115)
* add len method (2530b629)
* log config on start (083e253c)
* manage cache cli (4a475ed3)
* better handle existing image and incoming ones (still not perfect) (dd3d9a6e)
* get local working (4a37ecd5)
* handle existing images before watch (e00f9078)
* get watch working (0ee03db5)
* fs will need cleanup later (98ff1e03)
* start watch work (25c35ad2)
* start work on watch (0ca2b95d)
* print comment (0a5064aa)
* handle interrupts (9392d5b6)
* list keys and stop gabdger logging (8c28c0f8)
* update sigs (804d2cfd)
* version cli flags (db5db9a3)
* update deps (bd581680)
* rename constructor and check in-flight map (cef7da3a)
* reorder cli flags (db2f6ae8)
* remove file after popping (f50fa771)
* ignore badger dir (62623ebf)
* real asserts in TestPrettyPrintFileSizes (b07776e9)
* ignore skipcache (352e62f1)
* clean up debug code (0f1f41ed)
* check cache before adding to the queue (573ca64f)
* added broken files to cache (05144fc2)
* rm GetExistingFiles (f23d333c)
* cache contains (01f3cde2)
* getting it to run errors (0790a1f0)
* check if already exists (0e6c24ac)
* gut GetExistingFiles (4a142079)
* init map and print queue (7252f431)
* enable RemoveOriginals (25ee1c52)
* deps (6d75957e)
* get appa compiling (83b99b03)
* mess with FS return types (bb7e19de)
* wire up main so it at least compiles (b2aa807c)
* update queue sig for tests (b2874802)
* period (952d4ad5)
* queue notifications (055a0b4f)
* upsize queue for now, watch later (260387ca)
* worker only upsizes one image (41279748)
* badger and uilive (de50745e)
* rage in main (e429ed14)
* start breaking out daemon (632b4d98)
* start of cache (bd4d49c3)
* simplify args (76467a9f)
* condense code with watchDir (7880193d)
* watchdir with dedup (1116b928)
* watchDir with already exists logic (0b357e8a)
* only consider images that are not already upsized (b50ed031)
* AddImage (643b6fbc)
* move mkdir to fs pkg (7cdd5416)
* first crack at listing the correct files (02a4d7dd)
* watch dir (25416108)
* add support for watch events on App (022b9454)
* scheduler comments (f98cf90a)
* comments an removing old funcs (03dde3ca)
* keep track of elaready processed images (cd44a6b5)
* remove old middleware and add AlreadyUpsized (58a93e12)
* queue comments (6f582e78)
* use ref (03f702f1)
* new cmd output capture (f5ce2627)
* wire up main to the app (e13cd095)
* init main func for local (f8f1da90)
* Run() to wire it all up (1b72e283)
* queue > chan (582c358f)
* worker no longer relies on channels, just a loop call to NextImage() (135d6315)
* queue package name (a84658f5)
* return slice from existing files rather than chan (8d7cc4e1)
* remove zero files (acee32fa)
* reorganize packages (e7d792f5)
* rename struct fields in Queue (28ca8dc2)
* refactor queue funcs (5f1bdae9)
* deps (9bfd5b20)
* move common code to lib (9bf833e7)
* move common code to lib (f1dcfce9)
* atart auto work (45e8d61a)
* add api (c47536db)
* build a queue with a linked list (5b5bc77c)
* easier builds (55b9e9f0)
* use path.Entry (f23bc7a5)
* ongoing work for daemon mode (d93a9ca9)
* take path.Entry + pretty print logs (707dda57)
* dedup existing files (4bc1d72f)
* logging (45b7ac70)
* upgrade path lib (58c7e0d1)
* upgrade to new path.WatchDir (201b400c)
* enable prom (eba11bce)
* clean up stdout of cmd call (70fc18a9)
* use path rather than string (3d00f4de)
* get fs watcher working (3b62dd29)
* fs deps (48f9dd65)
* disable for now (ead4af8d)
* get static files served (e138fb4a)
* get ws working (840b267a)
* example curl (d1fc3358)
* get it working (148ee035)
* init (b523326a)
* Initial commit (6ddf7f51)




cmd/auto/main.go
package main

import (
	"flag"
	"net/http"
	"os"
	"os/signal"
	"path/filepath"
	"syscall"
	"time"

	"github.com/jaypipes/ghw"
	"github.com/kmulvey/path"
	"github.com/kmulvey/realesrgan-scheduler/internal/app/realesrgan/local"
	"github.com/kmulvey/realesrgan-scheduler/internal/fs"
	"github.com/kmulvey/realesrgan-scheduler/internal/pkg/ignoreregex"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	log "github.com/sirupsen/logrus"
	"go.szostok.io/version"
	"go.szostok.io/version/printer"
)

const promNamespace = "realesrgan_scheduler"

func main() {

	log.SetFormatter(&log.TextFormatter{
		FullTimestamp:   true,
		TimestampFormat: "2006-01-02 15:04:05",
	})

	go func() {
		http.Handle("/metrics", promhttp.Handler())
		s := &http.Server{
			Addr:           ":6000",
			ReadTimeout:    10 * time.Second,
			WriteTimeout:   10 * time.Second,
			MaxHeaderBytes: 1 << 20,
		}
		log.Fatal(s.ListenAndServe())
	}()

	// get the user options
	var originalImages, upscaledImages, cacheDir path.Entry
	var realesrganPath, skipFile string
	var daemon, removeOriginals, h, ver bool
	var numGPUs int

	flag.Var(&originalImages, "original-images-dir", "path to the original (input) images")
	flag.Var(&upscaledImages, "upscaled-images-dir", "where to store the upscaled images")
	flag.Var(&cacheDir, "cache-dir", "where to store the cache file for failed upsizes")
	flag.StringVar(&realesrganPath, "realesrgan-path", "realesrgan-ncnn-vulkan", "where the realesrgan binary is")
	flag.StringVar(&skipFile, "skip-file", "", "file with directories to skip, one per line")
	flag.IntVar(&numGPUs, "num-gpus", 1, "how many gpus to use")
	flag.BoolVar(&removeOriginals, "remove-originals", false, "delete original images after upsizing")
	flag.BoolVar(&daemon, "d", false, "run as a daemon (does not quit)")
	flag.BoolVar(&ver, "version", false, "print version")
	flag.BoolVar(&h, "help", false, "print options")
	flag.Parse()

	if h {
		flag.PrintDefaults()
		os.Exit(0)
	}

	if ver {
		var verPrinter = printer.New()
		var info = version.Get()
		if err := verPrinter.PrintInfo(os.Stdout, info); err != nil {
			log.Fatal(err)
		}
		os.Exit(0)
	}

	if gpus, err := getNumGPUs(); err != nil {
		log.Fatal("error getting gpu info: ", err)
	} else if numGPUs > gpus {
		log.Fatalf("cannot use %d gpus as there are only %d.", numGPUs, gpus)
	}

	log.Infof("Config: originalImages: %s, upscaledImages: %s, realesrganPath: %s, cacheDir: %s, skipFIle: %s, removeOriginals: %t, daemon: %t, numGpus: %d",
		originalImages.AbsolutePath,
		upscaledImages.AbsolutePath,
		realesrganPath,
		cacheDir.AbsolutePath,
		skipFile,
		removeOriginals,
		daemon,
		numGPUs)

	var upsizedDirs, err = path.List(upscaledImages.AbsolutePath, 2, false, path.NewDirEntitiesFilter())
	if err != nil {
		log.Fatalf("error getting existing upsized dirs: %s", err)
	}

	rl, err := local.NewRealesrganLocal(promNamespace, cacheDir.AbsolutePath, realesrganPath, upscaledImages.AbsolutePath, numGPUs, removeOriginals)
	if err != nil {
		log.Fatalf("error in: NewRealesrganLocal %s", err)
	}

	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	go func() {
		for range c {
			log.Info("interrupt caught, closing db ...")
			err = rl.Cache.Close()
			if err != nil {
				log.Fatalf("error closing badger: %s", err)
			}

			log.Info("bye")
			os.Exit(0)
		}
	}()

	// for each upsized directory, go back to its "originals" dir and look for additional files that have not been upsized.
	for _, upsizedDir := range upsizedDirs {

		var upsizedBase = filepath.Base(upsizedDir.AbsolutePath)
		var originalsDir = filepath.Join(originalImages.AbsolutePath, upsizedBase)

		var re, err = ignoreregex.SkipFileToRegexp(skipFile)
		if err != nil {
			log.Fatalf("error creating regex to skip dirs: %s", err)
		}

		if re.MatchString(originalsDir) {
			log.Infof("skipping %s ...", originalsDir)
			continue
		}

		rl.SetOutputPath(upsizedDir.AbsolutePath)

		originalImages, err := path.List(originalsDir, 2, false, path.NewRegexEntitiesFilter(fs.ImageExtensionRegex))
		if err != nil {
			log.Fatalf("error getting existing original images: %s", err)
		}

		if len(originalImages) == 0 {
			continue
		}

		log.Infof("Starting queue length: %d for dir: %s", rl.Queue.Len(), originalsDir)

		err = rl.Run(originalImages)
		if err != nil {
			log.Errorf("error in Run(): %s", err)
		}
	}
}

func getNumGPUs() (int, error) {

	var gpu, err = ghw.GPU()
	if err != nil {
		return 0, err
	}

	return len(gpu.GraphicsCards), nil
}
